python/cudf/cudf/core/indexing_utils.py

```diff
@@ -0,0 +1,225 @@
+    # No special-casing, key gets rows, and if a dataframe second part
+    # gets all columns
+    return [key, slice(None)][:n]
+
+
+def unpack_dataframe_iloc_indexer(
+    key: Any, frame: cudf.DataFrame
+) -> Tuple[Any, Tuple[bool, ColumnLabels]]:
+    """Unpack and index key for DataFrame iloc getitem.
+
+    Parameters
+    ----------
+    key
+        Key to unpack
+    frame
+        DataFrame for unpacking context
+
+    Returns
+    -------
+    tuple
+        2-tuple of a key for the rows and tuple of
+        (scalar_column_index, column_names) for the columns
+
+    Raises
+    ------
+    TypeError
+        If the column indexer is invalid
+    """
+    rows, cols = unpack_iloc_key(key, frame, len(frame.shape))
+    if cols is Ellipsis:
+        cols = slice(None)
+    scalar = is_integer(cols)
+    try:
+        column_names: ColumnLabels = list(
+            frame._data.get_labels_by_index(cols)
+        )
+        if len(set(column_names)) != len(column_names):
+            raise NotImplementedError(
+                "cudf DataFrames do not support repeated column names"
+            )
+    except TypeError:
+        raise TypeError(
+            "Column indices must be integers, slices, "
+            "or list-like of integers"
+        )
+    return (rows, (scalar, column_names))
+
+
+def unpack_series_iloc_indexer(key: Any, frame: cudf.Series) -> Any:
+    """Unpack an index key for Series iloc getitem.
+
+    Parameters
+    ----------
+    key
+        Key to unpack
+    frame
+        Series for unpacking context
+
+    Returns
+    -------
+    Single key that will index the rows
+    """
+    (rows,) = unpack_iloc_key(key, frame, len(frame.shape))
+    return rows
+
+
+def normalize_row_iloc_indexer(
+    key: Any, n: int, check_bounds=False
+) -> IndexSpec:
+    """
+    Normalize and produce structured information about a row indexer
+
+    Given a row indexer that has already been normalized by
+    :func:`unpack_iloc_key`, inspect further and produce structured
+    information for indexing operations to act upon.
+
+    Parameters
+    ----------
+    key
+        Suitably normalized key for row indexing
+    n
+        Length of frame to index
+    check_bounds
+        If True, perform bounds checking of the key if it is a gather
+        map.
+
+    Returns
+    -------
+    IndexSpec
+        Structured data for indexing. The first entry is a
+        :class:`Indexer` tag, the second entry is normalized
+        arguments to the tag-specific indexing routine.
+
+    Raises
+    ------
+    IndexError
+        If a valid type of indexer is provided, but it is out of
+        bounds
+    TypeError
+        If the indexing key is otherwise invalid.
+    """
+    if key is Ellipsis:
+        key = slice(None)
+    if isinstance(key, slice):
+        return (IndexTag.SLICE, key)
+    else:
+        if _is_scalar_or_zero_d_array(key):
+            key = np.asarray(key)
```

Comment:
If `key` happens to be a scalar/array on the device then we're copying it to host and back again to device later. Can we avoid that?

Reply:
Likely yes...

Reply:
I now just directly turn it into a column and then inspect the dtype there.